lib/spack/llnl/util/filesystem.py

```diff
@@ -1712,66 +1715,123 @@ def find(root, files, recursive=True):
+def find_max_depth(root, globs, max_depth: Optional[int] = None):
+    """Given a set of non-recursive glob file patterns, finds all
+    files matching those patterns up to a maximum specified depth.
 
-    # Make the path absolute to have os.walk also return an absolute path
-    root = os.path.abspath(root)
-    for path, _, list_files in os.walk(root):
-        for search_file in search_files:
-            matches = glob.glob(os.path.join(path, search_file))
-            matches = [os.path.join(path, x) for x in matches]
-            found_files[search_file].extend(matches)
+    If a directory has a name which matters an input pattern, it will
+    not be included in the results.
 
-    answer = []
-    for search_file in search_files:
-        answer.extend(found_files[search_file])
+    Does not search in directories below the specified depth.
 
-    return answer
+    If ``globs`` is a list, files matching earlier entries are placed
+    in the return value before files matching later entries.
+    """
+    if not os.path.exists(root):
+        return []
 
+    if max_depth is None:
+        max_depth = sys.maxsize
 
-@system_path_filter
-def _find_non_recursive(root, search_files):
-    # The variable here is **on purpose** a defaultdict as os.list_dir
-    # can return files in any order (does not preserve stability)
-    found_files = collections.defaultdict(list)
+    if isinstance(globs, str):
+        globs = [globs]
+    # Apply normcase to regular expressions and to the filenames:
+    # this respects case-sensitivity semantics of different OSes
+    # (e.g. file search is typically case-insensitive on Windows)
+    regexes = [re.compile(fnmatch.translate(os.path.normcase(x))) for x in globs]
 
-    # Make the path absolute to have absolute path returned
+    if not os.path.isdir(root):
+        raise ValueError(f"{root} is not a directory")
+
+    # Note later calls to os.scandir etc. return abspaths if the
+    # input is absolute, see https://docs.python.org/3/library/os.html#os.DirEntry.path
     root = os.path.abspath(root)
 
-    for search_file in search_files:
-        matches = glob.glob(os.path.join(root, search_file))
-        matches = [os.path.join(root, x) for x in matches]
-        found_files[search_file].extend(matches)
+    found_files = collections.defaultdict(list)
+
+    def _follow_link(p):
+        last = p
+        while os.path.islink(last):
+            last = resolve_link_target_relative_to_the_link(last)
+        return last
+
+    def _dir_id(p):
+        stat_result = os.stat(p, follow_symlinks=True)
+        # Note: on windows, st_ino is the file index and st_dev
+        # is the volume serial number. See
+        # https://github.com/python/cpython/blob/3.9/Python/fileutils.c
+        return (stat_result.st_ino, stat_result.st_dev)
+
+    visited_dirs = set([_dir_id(root)])
+
+    # Each queue item stores the depth and path
+    # BFS traversal is important to ensure repeatable results given
```

Comment:
"repeatable results" is a bit confusing :thinking: each unique directory is recursed into only once, so it must occur at lowest depth for `max_depth` to be respected. That applies when running find_max_depth once, so I would not talk about repeatable results.

Reply:
> "repeatable results" is a bit confusing 

"repeatable" means you get the same result every time for the same layout. If there's another word you want to use, I can use it, but without that, IMO it describes exactly what we want and what we get (and I want to use a word that is at least as informative as "repeatable").

> each unique directory is recursed into only once, so it must occur at lowest depth for max_depth to be respected. 

If you do DFS there's no guarantee you won't reach a given directory via a longer path (because we now follow symlinks to siblings of parents). In general though max_depth doesn't guarantee anything about repeatability, say we pick a max_depth of `10^6`, that will cover every dir structure we ever traversed (and in particular not distinguish between any of them)

> so I would not talk about repeatable results.

I think talking about repeatable results is important, in particular for tests. Without repeatability, the tests could start failing based on arbitrary changes to traversal order.

Reply:
There are as many BFS orderings as there are DFS orderings, so don't talk about "repeatable results" when it comes to choosing BFS in favor of DFS.

Again, the *only* reason to pick BFS over DFS is: we locate files on the lowest depth to respect `max_depth`.

You can deterministically pick an ordering (whether DFS or BFS) by fixing the order the children are enqueued or recursed, but you are not doing this as there is no determinism in `os.scandir(path)`.

So, if you want to explain why BFS is picked, mention it is to respect max_depth and nothing else...